### ISSUE TITLE
Can the pre-trained embedding be used for unsupervised tasks, e.g. clustering, outlier detection?

### ISSUE THREAD

**Issue by tim5go** [dreamquark-ai/tabnet#303]
As titled, is it possible to perform an unsupervised task purely based on the embedding learned from `TabNetPretrainer`?

**Comment by Optimox** [comment]
Hello @tim5go,

Short answer: no this is not available through the library.

Longer answer: 
- for outlier detection for example, as the definition of what an outlier is vague, you could pretrain a model in an unsupervised way and compute the reconstruction error for each sample. Samples with large errors might be consider as outliers (since they are not easily reconstructed).
- for clustering, I guess you would need to access the intermediate outputs of the encoder, this is not accessible directly with the library, however you can easily add a parameter to the forward method `decoder_output=True` which would output the desired layer. You'll then have some features on which you can apply any clustering algorithm.

**Comment by tim5go** [comment]
@Optimox 
I see. Thanks for your explanation! 
Just to make sure, when you are talking about intermediate outputs of the encoder, I think you're referring to `steps_out` in the following line, right?
https://github.com/dreamquark-ai/tabnet/blob/b6e1ebaf694f37ad40a6ba525aa016fd3cec15da/pytorch_tabnet/tab_network.py#L373

For the outlier detection part, I tried the reconstruction loss approach before. But the result on my dataset is not quite satisfactory (Even simple VAE produces better result.) That's why I'm looking for alternative.

**Comment by Optimox** [comment]
Yes exactly, you could some those steps to get only one vector.

I guess what you are trying to do is an edge case for tabnet, I would personally go for simple VAE as well or an IsolationForest if willing to do outlier detection.

**Comment by monk1337** [comment]
Hi, I am also trying unsupervised task clustering on my tabular data. I have mix types of data ( continuous data and categorical data) @tim5go did you find any success on clustering task?

- @Optimox How can I perform clustering on my tabular data, one way I am thinking : getting the representation of tabular data and sending to k-means? any other ideas?
- And how to get good representation of tabular data for clustering?
